### DISCUSSION TITLE
pan to top of image at top of viewport

### DISCUSSION THREAD

**Discussion by jrochkind** [openseadragon/openseadragon#2578] [Q&A]
i'd like to align the top of the image with the top of the viewport. 

(It is also at a zoom level where the entire image does not fit in viewport, not sure if that matters)

The various pan methods seem to have me supply what should be the _centerpoint_ of the image in the viewport after pan.  There of course should be some math I can do to find the centerpoint to achieve what I want taking into account current viewport size, current zoom level, image size, etc., but my head is swimming trying to figure it out!

To make things harder -- I'm about to zoom. I want to set it up so AFTER the zoom, the top of the image will be at top of viewport. So the _current_ zoom is not the right math, and built-in methods that assume the current zoom won't return the right results either. And I couldn't even figure it out before putting in that extra wrinkle!

Has anyone figured this out before, or can more easily than I?

**Comment by msalsbery** [comment]
@jrochkind

Everything is overridable to provide the behavior you want.

To understand coordinate systems in the viewer, see [Viewport  Coordinates](https://openseadragon.github.io/examples/viewport-coordinates/)

Zoom gets a little trickier since there's different ways to zoom (e.g. click to zoom, mouse scroll wheel, etc.). The [Viewer class](https://openseadragon.github.io/docs/OpenSeadragon.Viewer.html) has many events that you can subscribe to and customize zoom behavior in your event handlers.

It's kind of a big topic but it'll get easier once you dig in. What part(s) are you stuck on at the moment?

**Reply by jrochkind** [reply]
Thanks! I don't think I need to over-ride anything. 

I have an image I have just loaded. I have just initiated a (not yet completed) zoom operation on it, to zoom to a certain (known) level, call it Z. 

I think I want to call a `viewport.panTo`. I know how to call that method. I want to call it with some x, y, such that at zoom level Z, the image's top will be aligned with the viewport top, and the image will be horizontally centered. 

I think to find x, y, I need to do some arithmetic based on Z (known zoom level), and the viewport size.  

Useful methods in doing this arithmetic might include all the dimensional conversion methods on Viewport... `*ToImageCoordinates`, `*ToViewportCoordinates`, `*ToViewportRectangle`.  

I am having trouble figuring out how to do the arithmatic to calculate x and y to create a point to pass to `panTo`, such that at zoom Z the image top will be aligned with viewport top. 

Does this make sense?

**Comment by iangilman** [comment]
I always use `fitBounds` for this sort of thing. Just figure out the rectangle that you want to feature and it'll take you there. If you want to specify precisely what will appear in the viewer, do a `getBounds` first so you know the correct aspect ratio.

**Reply by jrochkind** [reply]
Thanks!

Exactly what I want rectangle I want to feature depends on what _fits_ at a given zoom level, such that the top of the image is aligned with top of the viewport.  There also could be space on the sides or even bottom depending on image size and zoom level. 

But those two methods may be helpful, I'll try to play around with them!
